### PR TITLE
Refactor CLI version selection

### LINKED ISSUE #575: No error shown when extension start up with invalid `cadence.flowCommand`
Is happening for me when configured to languageserver `run.sh`, may be related to specific functionality with this.

### LINKED ISSUE #583: Cannot change flow-cli binary when not in workspace/workspace folder

### PR DESCRIPTION
**Pull request by jribbink**
Closes #575 
Closes #583 

## Description

The code was maybe a bit cluttered/hard to follow.  This attempts to break the functionality up a bit better (moves caching into its own class).

Few functionality changes also:

- Hoists currently selected binary to top of list when choosing
- `cadence.changeCadenceVersion` command was a bit of a misnomer which came from the initial vision of this feature (but has deviated since).  This was changed to `cadence.changeFlowCliBinary`
- Also fixes a bug where the CLI version selection would break if an invalid flow command was configured at startup.
- Fixes "bug" when not in workspace folder

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: extension/src/dependency-installer/installers/flow-cli-installer.ts [hunk 1]
```diff
@@ -97,10 +98,10 @@ export class InstallFlowCLI extends Installer {
     }
   }
 
-  async checkVersion (vsn?: semver.SemVer): Promise<boolean> {
+  async checkVersion (vsn: semver.SemVer): Promise<boolean> {
     // Get user's version informaton
     this.#context.cliProvider.refresh()
-    const version = vsn ?? await this.#context.cliProvider.getAvailableBinaries().then(x => x.find(y => y.name === 'flow')?.version)
+    const version = vsn
```
Comment (chasefleming) [inline extension/src/dependency-installer/installers/flow-cli-installer.ts:104]: Do you need this line now?

File: extension/src/dependency-installer/installers/flow-cli-installer.ts [hunk 2]
```diff
@@ -128,7 +129,8 @@ export class InstallFlowCLI extends Installer {
   async verifyInstall (): Promise<boolean> {
     // Check if flow version is valid to verify install
     this.#context.cliProvider.refresh()
-    const version = await this.#context.cliProvider.getAvailableBinaries().then(x => x.find(y => y.name === 'flow')?.version)
+    const installedVersions = await this.#context.cliProvider.getBinaryVersions()
```
Comment (chasefleming) [inline extension/src/dependency-installer/installers/flow-cli-installer.ts:132]: Should you try/catch this? What if it fails?

File: extension/src/flow-cli/binary-versions-provider.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+const CHECK_FLOW_CLI_CMD = (flowCommand: string): string => `${flowCommand} version --output=json`
+const CHECK_FLOW_CLI_CMD_NO_JSON = (flowCommand: string): string => `${flowCommand} version`
+
+export enum KNOWN_FLOW_COMMANDS {
+  DEFAULT = 'flow',
+  CADENCE_V1 = 'flow-c1',
+}
+
+const LEGACY_VERSION_REGEXP = /Version:\s*(v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(\s|$)/m
```
Comment (chasefleming) [inline extension/src/flow-cli/binary-versions-provider.ts:15]: Can you add a comment for what this matches? It's pretty complicated.

File: extension/src/flow-cli/binary-versions-provider.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      return { command: bin, version }
+    } catch {
+      // Fallback to old method if JSON is not supported/fails
+      return await this.#fetchBinaryInformationOld(bin)
+    }
+  }
+
+  // Old version of fetchBinaryInformation (before JSON was supported)
+  // Used as fallback for old CLI versions
+  async #fetchBinaryInformationOld (bin: string): Promise<CliBinary | null> {
```
Comment (chasefleming) [inline extension/src/flow-cli/binary-versions-provider.ts:100]: Is there anyway to consolidate these two functions? Or to break out the common functionality into reusable methods (e.g. parsing version)

File: extension/src/flow-cli/binary-versions-provider.ts (hunk 3 above)
Comment (jribbink) [reply extension/src/flow-cli/binary-versions-provider.ts:15]: Yeah, will add.  Hopefully this won't be around much longer, it's kind of just a stopgap for now.

**Review by chasefleming — APPROVED** [review]

File: extension/src/flow-cli/cli-versions-provider.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+const CHECK_FLOW_CLI_CMD_NO_JSON = (flowCommand: string): string => `${flowCommand} version`
+
+export enum KNOWN_FLOW_COMMANDS {
+  DEFAULT = 'flow',
+  CADENCE_V1 = 'flow-c1',
+}
+
+// This regex matches a string like "Version: v{SEMVER}" and extracts the version number
+// It uses the official semver regex from https://semver.org/
+const LEGACY_VERSION_REGEXP = /Version:\s*(v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(\s|$)/m
```
Comment (chasefleming) [inline extension/src/flow-cli/cli-versions-provider.ts:17]: Does the `valid` method from the `semver` library support this format, so you could call `semver.valid` on the version instead of having to maintain this regex?

Comment (jribbink) [reply extension/src/flow-cli/cli-versions-provider.ts:17]: I can probably just change it to `/Version:\s*(.*)(\s|$)/m` & use `semver.parse`/`semver.valid` like you said, yeah
